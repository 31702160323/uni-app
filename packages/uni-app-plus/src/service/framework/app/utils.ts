import { extend } from '@vue/shared'
import { addLeadingSlash, parseQuery } from '@dcloudio/uni-shared'
import { createLaunchOptions, LaunchOptions } from '@dcloudio/uni-core'
import { $reLaunch } from '../../api/route/reLaunch'

export const EVENT_BACKBUTTON = 'backbutton'

export function backbuttonListener() {
  uni.navigateBack({
    from: 'backbutton',
    success() {}, // 传入空方法，避免返回Promise，因为onBackPress可能导致fail
  } as UniApp.NavigateBackOptions)
}

const enterOptions: LaunchOptions = createLaunchOptions()
const launchOptions: LaunchOptions = createLaunchOptions()

export function getLaunchOptions() {
  return launchOptions
}

export function getEnterOptions() {
  return enterOptions
}

export function initEnterOptions({
  path,
  query,
  referrerInfo,
}: Partial<RedirectInfo>) {
  extend(enterOptions, {
    path,
    query: query ? parseQuery(query) : {},
    referrerInfo: referrerInfo || {},
  })
}

export function initEnterReLaunch(info: RedirectInfo) {
  __uniConfig.realEntryPagePath =
    __uniConfig.realEntryPagePath || __uniConfig.entryPagePath
  __uniConfig.entryPagePath = info.path
  __uniConfig.entryPageQuery = info.query
  $reLaunch(
    { url: addLeadingSlash(info.path) + info.query },
    { resolve() {}, reject() {} }
  )
}

export function initLaunchOptions({
  path,
  query,
  referrerInfo,
}: Partial<RedirectInfo>) {
  extend(launchOptions, {
    path,
    query: query ? parseQuery(query) : {},
    referrerInfo: referrerInfo || {},
  })
  extend(enterOptions, launchOptions)
  return launchOptions
}

interface RedirectInfo extends Omit<LaunchOptions, 'query' | 'scene'> {
  query: string
  userAction: boolean
}

export function parseRedirectInfo(): RedirectInfo | void {
  const weexPlus = weex.requireModule('plus')
  if (weexPlus.getRedirectInfo) {
    const { path, query, extraData, userAction, fromAppid } =
      weexPlus.getRedirectInfo() || {}
    const referrerInfo: UniApp.UniConfig['referrerInfo'] = {
      appId: fromAppid,
      extraData: {},
    }
    if (extraData) {
      referrerInfo.extraData = extraData
    }
    return {
      path,
      query: query ? '?' + query : '',
      referrerInfo,
      userAction,
    }
  }
}
