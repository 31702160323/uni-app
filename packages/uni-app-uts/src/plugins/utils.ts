import path from 'path'
import { init, parse } from 'es-module-lexer'
import { normalizePath, removeExt } from '@dcloudio/uni-cli-shared'
import { camelize, capitalize } from '@vue/shared'

export const ENTRY_FILENAME = 'index.uts'

export async function parseImports(code: string) {
  await init
  const [imports] = parse(code)
  return imports
    .map(({ s, e }) => {
      return `import "${code.slice(s, e)}"`
    })
    .join('\n')
}

export function uvueOutDir() {
  return path.join(process.env.UNI_OUTPUT_DIR, '../.uvue')
}

export function genClassName(fileName: string) {
  return (
    'Gen' +
    capitalize(camelize(removeExt(normalizePath(fileName).replace(/\//g, '-'))))
  )
}

export function isVue(filename: string) {
  return filename.endsWith('.vue') || filename.endsWith('.uvue')
}

export function stringifyMap(obj: unknown) {
  return mapToInitString(objToMap(obj as Record<string, unknown>), true)
}

function mapToInitString(
  map: Map<string, unknown>,
  ts: boolean = false
): string {
  let entries = []
  for (let [key, value] of map) {
    if (value instanceof Map) {
      entries.push(`["${key}", ${mapToInitString(value, ts)}]`)
    } else {
      entries.push(`["${key}", ${JSON.stringify(value)}]`)
    }
  }
  return `new Map${ts ? '<string, any>' : ''}([${entries.join(', ')}])`
}

function objToMap(obj: Record<string, unknown>) {
  const map = new Map()
  for (const key in obj) {
    const value = obj[key]
    if (typeof value === 'object') {
      map.set(key, objToMap(value as Record<string, unknown>))
    } else {
      map.set(key, value)
    }
  }
  return map
}
