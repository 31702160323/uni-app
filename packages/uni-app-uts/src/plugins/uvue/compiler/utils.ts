import { CompilerOptions } from './options'

export function genRenderFunctionDecl({
  targetLanguage,
  filename,
}: CompilerOptions): string {
  return `${
    targetLanguage === 'kotlin' ? '@Suppress("UNUSED_PARAMETER") ' : ''
  }function ${filename}Render(_ctx: ${filename}): VNode | null`
}
