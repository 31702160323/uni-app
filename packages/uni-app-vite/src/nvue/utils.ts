export const external = ['vue', 'vuex']
export const globals = {
  vue: 'Vue',
  vuex: 'Vuex',
  pinia: 'Pinia',
}
export const esbuildGlobals = {
  vue: 'Vue',
  vuex: 'uni.Vuex',
  pinia: 'uni.Pinia',
}
