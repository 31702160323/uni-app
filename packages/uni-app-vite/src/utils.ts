import path from 'path'
export const templateDir = path.resolve(__dirname, '../lib/template/')
export function nvueOutDir(appService: boolean) {
  return path.join(
    process.env.UNI_OUTPUT_DIR,
    '../.nvue' + (appService ? '/.app-service' : '')
  )
}
