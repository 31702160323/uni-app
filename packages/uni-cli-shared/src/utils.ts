import fs from 'fs'
import os from 'os'
import path from 'path'
import { camelize, capitalize } from '@vue/shared'
import { once } from '@dcloudio/uni-shared'
export { default as hash } from 'hash-sum'

import { PAGE_EXTNAME, PAGE_EXTNAME_APP } from './constants'
import { SFCTemplateCompileOptions } from '@vue/compiler-sfc'
import { NodeTypes, ElementNode, Node } from '@vue/compiler-core'
export const isWindows = os.platform() === 'win32'
export function normalizePath(id: string): string {
  return isWindows ? id.replace(/\\/g, '/') : id
}

export function checkElementNodeTag(
  node: Node | null | undefined,
  tag: string
): node is ElementNode {
  return (
    !!node &&
    node.type === NodeTypes.ELEMENT &&
    (node as ElementNode).tag === tag
  )
}

export const resolveMainPathOnce = once((inputDir: string) => {
  const mainTsPath = path.resolve(inputDir, 'main.ts')
  if (fs.existsSync(mainTsPath)) {
    return normalizePath(mainTsPath)
  }
  return normalizePath(path.resolve(inputDir, 'main.js'))
})

export function resolveBuiltIn(path: string) {
  if (process.env.UNI_CLI_CONTEXT) {
    return require.resolve(path, { paths: [process.env.UNI_CLI_CONTEXT] })
  }
  return require.resolve(path)
}

export function normalizeIdentifier(str: string) {
  return capitalize(camelize(str.replace(/\//g, '-')))
}

export function normalizePagePath(pagePath: string, platform: UniApp.PLATFORM) {
  const absoltePagePath = path.resolve(process.env.UNI_INPUT_DIR, pagePath)
  let extnames = PAGE_EXTNAME
  if (platform === 'app') {
    extnames = PAGE_EXTNAME_APP
  }
  for (let i = 0; i < extnames.length; i++) {
    const extname = extnames[i]
    if (fs.existsSync(absoltePagePath + extname)) {
      return pagePath + extname
    }
  }
  console.error(`${pagePath} not found`)
}

export function removeExt(str: string) {
  return str.split('?')[0].replace(/\.\w+$/g, '')
}

const NODE_MODULES_REGEX = /(\.\.\/)?node_modules/g

export function normalizeNodeModules(str: string) {
  str = normalizePath(str).replace(NODE_MODULES_REGEX, 'node-modules')
  // HBuilderX 内置模块路径转换
  str = str.replace(/.*\/plugins\/uniapp-cli\/node[-_]modules/, 'node-modules')
  if (process.env.UNI_PLATFORM === 'mp-alipay') {
    str = str.replace('node-modules/@', 'node-modules/npm-scope-')
  }
  return str
}

export function normalizeMiniProgramFilename(
  filename: string,
  inputDir?: string
) {
  if (!inputDir || !path.isAbsolute(filename)) {
    return normalizeNodeModules(filename)
  }
  return normalizeNodeModules(path.relative(inputDir, filename))
}

export function createUniVueTransformAssetUrls(
  base: string
): SFCTemplateCompileOptions['transformAssetUrls'] {
  return {
    base,
    tags: {
      audio: ['src'],
      video: ['src', 'poster'],
      img: ['src'],
      image: ['src'],
      'cover-image': ['src'],
      // h5
      'v-uni-audio': ['src'],
      'v-uni-video': ['src', 'poster'],
      'v-uni-image': ['src'],
      'v-uni-cover-image': ['src'],
      // nvue
      'u-image': ['src'],
      'u-video': ['src', 'poster'],
    },
  }
}
