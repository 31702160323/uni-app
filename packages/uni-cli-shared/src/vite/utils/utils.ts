import type { ResolvedConfig, UserConfig } from 'vite'

export function withSourcemap(config: ResolvedConfig) {
  if (config.command === 'serve') {
    return true
  }
  return !!config.build.sourcemap
}

export function isInHybridNVue(config: UserConfig | ResolvedConfig): boolean {
  return (config as any).nvue && process.env.UNI_RENDERER !== 'native'
}
