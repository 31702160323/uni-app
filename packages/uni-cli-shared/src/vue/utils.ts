import { isComponentTag } from '@dcloudio/uni-shared'
import {
  ComponentNode,
  ElementTypes,
  isCoreComponent,
  NodeTypes,
  RootNode,
  TemplateChildNode,
  TransformContext,
} from '@vue/compiler-core'

export function isUserComponent(
  node: RootNode | TemplateChildNode,
  context: TransformContext
): node is ComponentNode {
  return (
    node.type === NodeTypes.ELEMENT &&
    node.tagType === ElementTypes.COMPONENT &&
    !isComponentTag(node.tag) &&
    !isCoreComponent(node.tag) &&
    !context.isBuiltInComponent(node.tag)
  )
}
