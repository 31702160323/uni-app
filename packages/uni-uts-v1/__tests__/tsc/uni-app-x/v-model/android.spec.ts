import { runAndroid } from '../../utils'

runAndroid('v-model')
