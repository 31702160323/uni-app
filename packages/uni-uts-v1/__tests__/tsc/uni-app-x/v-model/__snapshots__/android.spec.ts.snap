// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`android v-model 1`] = `
"import { defineComponent, useModel } from 'vue';
defineComponent({
    props: {
        modelValue: { type: String },
        msg: { type: String, default: 'default msg' },
        num: { type: Number, default: 1 },
    },
    setup(props) {
        const modelValue = useModel(props, 'modelValue');
        modelValue.value = '123';
        const msg = useModel(props, 'msg');
        msg.value = '123';
        const num = useModel(props, 'num');
        num.value = 1;
    },
});
"
`;
